jaxtyping/_import_hook.py

```diff
@@ -88,10 +90,58 @@ def _dot_lookup(*elements):
     return out
 
 
-def _str_lookup(string):
-    module = ast.parse(string)
-    (expr,) = module.body
-    return expr.value
+def _get_hash(string):
+    if string is None:
+        return 0
+    if callable(string):
+        return _get_hash(getsource(string))
+    return str(hashlib.md5(string.encode("utf-8")).hexdigest())
+
+
+_decorator_lookup = {}
+_decorator_ast_lookup = {}
+
+
+def _typechecker_to_ast(typechecker):
+    hash = _update_typechecker_decorator(typechecker)
+    if hash not in _decorator_ast_lookup.keys():
+        _decorator_ast_lookup[hash] = (
+            ast.parse(
+                f"@jaxtyping._import_hook._decorator_lookup['{hash}']\n"  # noqa: E501
+                "def _():\n    ..."
+            )
+            .body[0]
+            .decorator_list[0]
+        )
+
+    return _decorator_ast_lookup[hash]
+
+
+def _update_typechecker_decorator(typechecker):
+    hash = _get_hash(typechecker)
+    if hash in _decorator_lookup.keys():
+        return hash
+
+    if type(typechecker) == str:
+        # If the typechecker is a string, we make a decorator ourselves
+        # and pass it to lookup
+        string_to_eval = (
+            "def f(x, *args, **kwargs):\n"
+            + f"  import {typechecker.split('.', 1)[0]}\n"
+            + f"  return {typechecker}(x, *args, **kwargs)"
+        )
+        local_vars = {}
+        exec(string_to_eval, {}, local_vars)
+        _decorator_lookup[hash] = local_vars["f"]
+    elif callable(typechecker):
+        # If the typechecker is a function, treat it as a decorator
+        _decorator_lookup[hash] = typechecker
+    else:
+        # If it is None, ignore it silently (use dummy decorator)
+        # If it is anything else, we still want to ignore it.
```

Comment:
I think I'd suggest raising an error in this case.

Reply:
Yep, makes sense. Thanks!